### DISCUSSION TITLE
pyscript with logging?

### DISCUSSION THREAD

**Discussion by csonsino** [python-cmd2/cmd2#1107] [Q&A]
I'm trying to figure out how to get the stdout text that was written through the python logging library while running a pyscript.

I have the following simple pyscript (`test_script.py`):
```
print("*** About to do push-ups...")
result = app("do_with_logger push-ups")
print("*** RESULT: " + str(result))

print("*** About to do sit-ups...")
result = app("do_with_print sit-ups")
print("*** RESULT: " + str(result))

app("quit")
```

Running through the following simple shell (`main.py`):
```
import cmd2
import logging


class CustomShell(cmd2.Cmd):
    prompt = "(customshell) "

    def __init__(self):
        super().__init__()

        logging.basicConfig(level=logging.INFO)
        # logging.basicConfig(level=logging.DEBUG, stream=self.stdout)
        self._logger = logging.getLogger("customlogger")

    do_parser = cmd2.Cmd2ArgumentParser()
    do_parser.add_argument("what", help="What to do")

    @cmd2.with_argparser(do_parser)
    def do_do_with_logger(self, args):
        msg = "doing " + str(args.what) + " ..."
        self._logger.info(msg)

    @cmd2.with_argparser(do_parser)
    def do_do_with_print(self, args):
        msg = "doing " + str(args.what) + " ..."
        print(msg)


if __name__ == '__main__':
    cshell = CustomShell()
    cshell.cmdloop()
```

If I run the pyscript with `python3 main.py "run_pyscript test_script.py"`, I get:
```
*** About to do push-ups...
INFO:customlogger:doing push-ups ...
*** RESULT: CommandResult(stdout='', stderr=None, stop=False, data=None)
*** About to do sit-ups...
*** RESULT: CommandResult(stdout='doing sit-ups ...\n', stderr=None, stop=False, data=None)
```

You can see that the logger appeared to write to `stdout` but the CommandResult did not capture it, whereas the print() wrote to `stdout` and it was captured in the CommandResult.

The main reason that I'm using the logging library is to be able to easily control the verbosity and amount of data that I get for each command.

**Comment by csonsino** [comment]
I just noticed the "Discussions" feature - let me know if you want me to move this over there instead of an Issue.

**Comment by tleonhardt** [comment]
@csonsino First off, if you are ever using `print` within a `cmd2` application, you are doing it wrong.

Please see the `cmd2` documentation on [Generating Output](https://cmd2.readthedocs.io/en/latest/features/generating_output.html)

`cmd2` has various built-in methods which generate output in a fashion that is aware of whether or not output should be redirected and if it is running within pyscript.
